clients/drcachesim/tools/filter/record_filter.cpp

```diff
@@ -105,35 +109,62 @@ record_filter_t::parallel_shard_supported()
     return true;
 }
 
-std::unique_ptr<std::ostream>
+std::string
 record_filter_t::get_writer(per_shard_t *per_shard, memtrace_stream_t *shard_stream)
 {
     per_shard->output_path = output_dir_ + DIRSEP + shard_stream->get_stream_name();
 #ifdef HAS_ZLIB
     if (ends_with(per_shard->output_path, ".gz")) {
         VPRINT(this, 3, "Using the gzip writer for %s\n", per_shard->output_path.c_str());
-        return std::unique_ptr<std::ostream>(new gzip_ostream_t(per_shard->output_path));
+        per_shard->file_writer =
+            std::unique_ptr<std::ostream>(new gzip_ostream_t(per_shard->output_path));
+        per_shard->writer = per_shard->file_writer.get();
+        return "";
+    }
+#endif
+#ifdef HAS_ZIP
+    if (ends_with(per_shard->output_path, ".zip")) {
+        VPRINT(this, 3, "Using the zip writer for %s\n", per_shard->output_path.c_str());
+        per_shard->archive_writer = std::unique_ptr<archive_ostream_t>(
+            new zipfile_ostream_t(per_shard->output_path));
+        per_shard->writer = per_shard->archive_writer.get();
+        return open_new_chunk(per_shard);
     }
 #endif
     VPRINT(this, 3, "Using the default writer for %s\n", per_shard->output_path.c_str());
-    return std::unique_ptr<std::ostream>(
+    per_shard->file_writer = std::unique_ptr<std::ostream>(
         new std::ofstream(per_shard->output_path, std::ofstream::binary));
+    per_shard->writer = per_shard->file_writer.get();
+    return "";
+}
+
+std::string
+record_filter_t::open_new_chunk(per_shard_t *shard)
+{
+    std::ostringstream stream;
+    stream << TRACE_CHUNK_PREFIX << std::setfill('0') << std::setw(4)
+           << shard->chunk_ordinal;
+    return shard->archive_writer->open_new_component(stream.str());
 }
 
 void *
 record_filter_t::parallel_shard_init_stream(int shard_index, void *worker_data,
                                             memtrace_stream_t *shard_stream)
 {
     auto per_shard = new per_shard_t;
-    per_shard->writer = get_writer(per_shard, shard_stream);
+    std::string error = get_writer(per_shard, shard_stream);
+    if (!error.empty()) {
+        per_shard->error = "Failure in opening writer: " + error;
+        success_ = false;
+    }
+    if (!per_shard->writer && !per_shard->archive_writer) {
```

Comment:
We probably want to check only per_shard->writer here.

Reply:
Done.